# Homework3

## Q1.1 
Clean up the dataset and create x_train, y_train, x_test, y_test in numpy arrays <br>
Provide summary statistics for the training and test datasets

In [1]:
import numpy as np
import pandas as pd

In [99]:
df_train = pd.read_csv('train.csv', header = None)
df_test = pd.read_csv('test.csv', header = None)

#### Remove rows with missing values

In [100]:
for i in range(15):
    df_train = df_train[df_train[i] != ' ?']
df_train = df_train.reset_index(drop=True)

for i in range(15):
    df_test = df_test[df_test[i] != ' ?']
df_test = df_test.reset_index(drop=True)

/anaconda3/lib/python3.7/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


#### Create x, y data 

In [101]:
y_train =  df_train[14]
y_train = y_train.to_frame()
x_train = df_train.copy()
del x_train[14]

In [102]:
x_train

0                  1       2            3   4                    5   \
0      39          State-gov   77516    Bachelors  13        Never-married   
1      50   Self-emp-not-inc   83311    Bachelors  13   Married-civ-spouse   
2      38            Private  215646      HS-grad   9             Divorced   
3      53            Private  234721         11th   7   Married-civ-spouse   
4      28            Private  338409    Bachelors  13   Married-civ-spouse   
...    ..                ...     ...          ...  ..                  ...   
30157  27            Private  257302   Assoc-acdm  12   Married-civ-spouse   
30158  40            Private  154374      HS-grad   9   Married-civ-spouse   
30159  58            Private  151910      HS-grad   9              Widowed   
30160  22            Private  201490      HS-grad   9        Never-married   
30161  52       Self-emp-inc  287927      HS-grad   9   Married-civ-spouse   

                       6               7       8        9      10  11  12  \
0            Adm-clerical   Not-in-family   White     Male   2174   0  40   
1         Exec-managerial         Husband   White     Male      0   0  13   
2       Handlers-cleaners   Not-in-family   White     Male      0   0  40   
3       Handlers-cleaners         Husband   Black     Male      0   0  40   
4          Prof-specialty            Wife   Black   Female      0   0  40   
...                   ...             ...     ...      ...    ...  ..  ..   
30157        Tech-support            Wife   White   Female      0   0  38   
30158   Machine-op-inspct         Husband   White     Male      0   0  40   
30159        Adm-clerical       Unmarried   White   Female      0   0  40   
30160        Adm-clerical       Own-child   White     Male      0   0  20   
30161     Exec-managerial            Wife   White   Female  15024   0  40   

                   13  
0       United-States  
1       United-States  
2       United-States  
3       United-States  
4                Cuba  
...               ...  
30157   United-States  
30158   United-States  
30159   United-States  
30160   United-States  
30161   United-States  

[30162 rows x 14 columns]

In [103]:
y_test =  df_test[14]
y_test = y_test.to_frame()
x_test = df_test.copy()
del x_test[14]

In [104]:
x_test

0              1       2              3   4                    5   \
0      25        Private  226802           11th   7        Never-married   
1      38        Private   89814        HS-grad   9   Married-civ-spouse   
2      28      Local-gov  336951     Assoc-acdm  12   Married-civ-spouse   
3      44        Private  160323   Some-college  10   Married-civ-spouse   
4      34        Private  198693           10th   6        Never-married   
...    ..            ...     ...            ...  ..                  ...   
15055  33        Private  245211      Bachelors  13        Never-married   
15056  39        Private  215419      Bachelors  13             Divorced   
15057  38        Private  374983      Bachelors  13   Married-civ-spouse   
15058  44        Private   83891      Bachelors  13             Divorced   
15059  35   Self-emp-inc  182148      Bachelors  13   Married-civ-spouse   

                       6               7                    8        9     10  \
0       Machine-op-inspct       Own-child                Black     Male     0   
1         Farming-fishing         Husband                White     Male     0   
2         Protective-serv         Husband                White     Male     0   
3       Machine-op-inspct         Husband                Black     Male  7688   
4           Other-service   Not-in-family                White     Male     0   
...                   ...             ...                  ...      ...   ...   
15055      Prof-specialty       Own-child                White     Male     0   
15056      Prof-specialty   Not-in-family                White   Female     0   
15057      Prof-specialty         Husband                White     Male     0   
15058        Adm-clerical       Own-child   Asian-Pac-Islander     Male  5455   
15059     Exec-managerial         Husband                White     Male     0   

       11  12              13  
0       0  40   United-States  
1       0  50   United-States  
2       0  40   United-States  
3       0  40   United-States  
4       0  30   United-States  
...    ..  ..             ...  
15055   0  40   United-States  
15056   0  36   United-States  
15057   0  50   United-States  
15058   0  40   United-States  
15059   0  60   United-States  

[15060 rows x 14 columns]

In [105]:
x_train.columns = ['age', 'workclass', 'fnlwgt', 'education', 'education-num', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week', 'native-country']
x_test.columns = ['age', 'workclass', 'fnlwgt', 'education', 'education-num', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week', 'native-country']
binary_col = ['workclass', 'education', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'native-country']

#### 1-of-K encoding
Only keep values that occur more than 10 times

In [106]:
for i in binary_col:
    df2 = pd.get_dummies(x_train[i])
    df2_col = df2.columns
    for j in df2_col:
        df2[j] = df2[j].astype('int8')
        sum_col = df2[j].sum()
        if sum_col < 10:
            del df2[j]
    x_train = x_train.join(df2)
    del x_train[i]

In [107]:
train_col = x_train.columns
for i in binary_col:
    df2 = pd.get_dummies(x_test[i])
    x_test = x_test.join(df2)
    del x_test[i]
test_col = x_test.columns
for i in test_col:
    if i not in train_col:
        del x_test[i]

In [108]:
y_train = y_train.replace(' <=50K', 0)
y_train = y_train.replace(' >50K', 1)
y_test = y_test.replace(' <=50K.', 0)
y_test = y_test.replace(' >50K.', 1)

In [109]:
print("x_train has", x_train.shape[0], "observations and", x_train.shape[1], "columns")
continuous_col = ['age', 'fnlwgt', 'education-num', 'capital-gain', 'capital-loss', 'hours-per-week']
print("summary of continuous features")
print(x_train[continuous_col].describe())
print("x_test has", x_test.shape[0], "observations and", x_test.shape[1], "columns")
print("summary of continuous features")
print(x_test[continuous_col].describe())

x_train has 30162 observations and 102 columns
summary of continuous features
                age        fnlwgt  education-num  capital-gain  capital-loss  \
count  30162.000000  3.016200e+04   30162.000000  30162.000000  30162.000000   
mean      38.437902  1.897938e+05      10.121312   1092.007858     88.372489   
std       13.134665  1.056530e+05       2.549995   7406.346497    404.298370   
min       17.000000  1.376900e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.176272e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.784250e+05      10.000000      0.000000      0.000000   
75%       47.000000  2.376285e+05      13.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours-per-week  
count    30162.000000  
mean        40.931238  
std         11.979984  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max      

In [110]:
# x_col = x_train.columns
# for i in x_col:
#     if i not in continuous_col:
#         print(x_train[i].value_counts())

In [111]:
x_train = x_train.values
y_train = y_train.values
x_test = x_test.values
y_test = y_test.values

## Q1.2
Derive the gradient and hessian matrix for the new E(w)

Our new error function $E(w) = (1/2)w^T\Lambda w - \Sigma_{i=1}^{n}[t_i\ln{y_i}+(1-t_i)\ln({1-y_i})]$ <br>
Gradient $\nabla E(w) = \Lambda w - \Sigma_{i=1}^{n}(t_i - y_i)x_i$ <br>
$\nabla E(w) = \Lambda w - \Sigma_{i=1}^{n}(t_i - \sigma(w^Tx_i))x_i = \Lambda w - \Sigma_{i=1}^{n}(t_i x_i - \sigma(w^T x_i)x_i)$ <br>
Hessian $H(w) = \Lambda + \Sigma_{i=1}^{n} (\sigma(w^T x_i)(1-\sigma(w^T x_i)x_i)*x_i) = \Lambda + \Sigma_{i=1}^{n} (y_i (1-y_i) x_i x_i) = \Lambda + \Sigma_{i=1}^{n} x_i^Ty_i (1-y_i) x_i$

## Q1.3
Create a mylogistic_l2 class <br>

In [178]:
def sigmoid(function):
    return 1/(1+np.exp(-function))
    
class mylogistic_l2:
    def __init__(self, reg_vec, max_iter, tol, add_intercept):
        self.reg_vec = reg_vec
        self.max_iter = max_iter
        self.tol = tol
        self.add_intercept = add_intercept
        self.weight = None
    
    def error(self, y_train, y_pred):
        lambda_weight= (1/2)*self.weight.T.dot(self.reg_lambda).dot(self.weight)
        loss = np.sum(y_train*(np.log(y_pred)) + (1-y_train)*(np.log(1-y_pred)))
        error = lambda_weight - loss
        return error
    
    def step(self, x_train, y_train, y_pred):
        lambda_weight = np.matmul(self.reg_lambda, self.weight)
        error = np.matmul(x_train.T, y_train - y_pred)
        gradient = lambda_weight - error
    
        matrix = np.zeros((len(y_pred), len(y_pred)))
        np.fill_diagonal(matrix, y_pred*(1-y_pred))
        hessian = self.reg_lambda + x_train.T.dot(matrix).dot(x_train)
        
        self.weight = self.weight - np.linalg.inv(hessian).dot(gradient)
                      
    def fit(self, x_train, y_train):
        if self.add_intercept:
            intercept = np.ones((x_train.shape[0],1))
            x_train = np.hstack((x_train, intercept))
        
        self.reg_lambda = np.zeros((len(self.reg_vec), len(self.reg_vec)))
        np.fill_diagonal(self.reg_lambda, self.reg_vec)
        
        b = np.mean(self.reg_vec)
        tmp = np.add(np.dot(x_train.T, x_train), b*np.identity(x_train.shape[1]))
        self.weight = np.dot(np.dot(np.linalg.inv(tmp), x_train.T), y_train)
        
        previous_grad = 0
        for i in range(self.max_iter):
            prob = np.dot(x_train, self.weight)
            prediction = sigmoid(prob)
            
            prev_error = self.error(y_train, prediction)
            self.step(x_train, y_train, prediction)
            
            prob = np.dot(x_train, self.weight)
            prediction = sigmoid(prob)
            error = self.error(y_train, prediction)
            
            if prev_error-error < self.tol:
                break
                      
    def predict(self, x_test):
        if self.add_intercept:
            intercept = np.ones((x_test.shape[0],1))
            x_test = np.hstack((x_test, intercept))
        prob = np.dot(x_test, self.weight)
        prediction = sigmoid(prob)
        prediction[prediction >= 0.5] = 1
        prediction[prediction < 0.5] = 0
        return prediction
    def show_weight(self):
        return self.weight

Show the learned  𝑤  as well as test accuracy for the cases below <br>

In [179]:
def accuracy(y_test, y_pred):
    accuracy = 0
    for i in range(len(y_test)):
        if y_test[i] == y_pred[i]:
            accuracy+=1
    return accuracy/len(y_test)

1. Case 1: lambda = 1 for all coefficients

In [180]:
lambda_1 = np.ones((x_train.shape[1]+1,1))
logic_1 = mylogistic_l2(reg_vec = lambda_1, max_iter = 1000, tol = 1e-5, add_intercept = True)
logic_1.fit(x_train, y_train)
ypred_1 = logic_1.predict(x_test)

In [181]:
print("test accuracy:", accuracy(y_test, ypred_1))
print("learned weight:")
logic_1.show_weight()

test accuracy: 0.8480743691899071
learned weight:


array([[ 2.48591297e-02],
       [ 7.26281835e-07],
       [ 1.85806163e-01],
       [ 3.16590245e-04],
       [ 6.38679434e-04],
       [ 2.90215841e-02],
       [ 1.89694411e-01],
       [-4.97719501e-01],
       [-3.10975202e-01],
       [-1.29181491e-01],
       [-7.94111556e-01],
       [-6.18745003e-01],
       [-1.17270706e+00],
       [-4.52249247e-01],
       [-5.44907403e-01],
       [-3.82721832e-01],
       [-9.99810289e-02],
       [-2.62922225e-01],
       [-6.29791388e-01],
       [-4.96565839e-01],
       [-3.08326937e-01],
       [-1.27584811e-01],
       [ 1.28534378e-01],
       [ 5.87953355e-01],
       [-2.50481306e-01],
       [ 2.99991415e-01],
       [-1.37579993e+00],
       [ 6.82097029e-01],
       [-1.00989638e-01],
       [-1.00975740e+00],
       [ 1.19612258e+00],
       [ 8.38966933e-01],
       [-9.56673066e-01],
       [-1.50523218e+00],
       [-1.08232391e+00],
       [-8.14848363e-01],
       [-8.78492150e-02],
       [-2.46913576e-02],
       [ 7.1

2. Case 2: lambda = 1 for all but the intercept, no regularization for intercept term.

In [182]:
lambda_2 = np.ones((x_train.shape[1],1))
no_reg = np.zeros((1,1))
lambda_2 = np.append(lambda_2, no_reg, axis = 0)
logic_2 = mylogistic_l2(reg_vec = lambda_2, max_iter = 1000, tol = 1e-5, add_intercept = True)
logic_2.fit(x_train, y_train)
ypred_2 = logic_2.predict(x_test)

In [183]:
print("test accuracy:", accuracy(y_test, ypred_2))
print("learned weight:")
logic_2.show_weight()

test accuracy: 0.847808764940239
learned weight:


array([[ 2.54336822e-02],
       [ 7.50706804e-07],
       [ 2.95324923e-01],
       [ 3.17024478e-04],
       [ 6.39652126e-04],
       [ 2.94914512e-02],
       [ 7.05717318e-01],
       [ 1.78026502e-02],
       [ 2.09126595e-01],
       [ 3.82731145e-01],
       [-2.79821710e-01],
       [-1.04552949e-01],
       [-9.31003049e-01],
       [ 9.08118759e-02],
       [-1.06210390e-01],
       [-5.68864876e-02],
       [ 7.06129017e-01],
       [ 5.36014978e-01],
       [ 1.16361326e-01],
       [ 1.37527872e-01],
       [-4.01951495e-01],
       [-1.13370201e-01],
       [-7.43801938e-02],
       [ 6.79669927e-02],
       [-1.95742295e-02],
       [-1.15929725e-02],
       [-1.15969699e+00],
       [ 2.67049827e-01],
       [ 2.18010722e-02],
       [-5.26538170e-01],
       [ 1.61452756e+00],
       [ 1.36750998e+00],
       [-4.92456773e-01],
       [-1.01532721e+00],
       [-6.05766406e-01],
       [-3.41948994e-01],
       [ 1.64109182e-01],
       [ 2.28422252e-01],
       [ 9.6

3. Case 3: lambda = 1 for numerical-valued features, lambda = 0.5 for binary-valued features, no regularization for intercept term.

In [184]:
binary_cnt = 102-6
numerical = np.ones((6,1))
binary = np.full((binary_cnt,1), 0.5)
no_reg = np.zeros((1,1))
lambda_3 = np.append(numerical, binary, axis = 0)
lambda_3 = np.append(lambda_3, no_reg, axis = 0)

logic_3 = mylogistic_l2(reg_vec = lambda_3, max_iter = 1000, tol = 1e-5, add_intercept = True)
logic_3.fit(x_train, y_train)
ypred_3 = logic_3.predict(x_test)

In [185]:
print("test accuracy:", accuracy(y_test, ypred_3))
print("learned weight:")
logic_3.show_weight()

test accuracy: 0.847675962815405
learned weight:


array([[ 2.54757306e-02],
       [ 7.51944312e-07],
       [ 3.19092342e-01],
       [ 3.17319913e-04],
       [ 6.40115459e-04],
       [ 2.95135990e-02],
       [ 7.66990116e-01],
       [ 7.62915468e-02],
       [ 2.68338059e-01],
       [ 4.43128006e-01],
       [-2.20821080e-01],
       [-4.63422171e-02],
       [-1.28758443e+00],
       [ 2.18992474e-01],
       [-2.26138746e-03],
       [ 2.18816240e-02],
       [ 9.74981438e-01],
       [ 7.54034099e-01],
       [ 2.92224729e-01],
       [ 2.93070077e-01],
       [-4.19612079e-01],
       [-1.04346489e-01],
       [-1.12963414e-01],
       [-3.78240639e-02],
       [ 3.76517833e-02],
       [-7.30516160e-02],
       [-2.08394382e+00],
       [ 1.86338344e-01],
       [ 5.48283060e-02],
       [-5.72016449e-01],
       [ 1.82571092e+00],
       [ 1.39641685e+00],
       [-5.47067011e-01],
       [-1.05936560e+00],
       [-6.55642137e-01],
       [-3.88036577e-01],
       [ 2.36234437e-01],
       [ 3.00253515e-01],
       [ 1.0

## Q1.4
Split the training data into subtraining (90%) and tuning (10%) to search for the best hyperparameters <br>
Set the regularization coefficient for the constant term to zero. Allow different regularizations for continuous-valued and binary-valued features <br>
Let  𝑎1  and  𝑎2  denote the regularization coefficients for continuous-valued and binary-valued features <br>
Search the best  𝑎1  and  𝑎2  and report the test accuracy using the best hyper-parameters. You should follow the following procedure to search for the best hyperparameters <br>

1. Choose a set of grids among a reasonable range. For example, 10 grids in [0.01, 100] 
2. Conduct grid search with the constraint that  𝑎1=𝑎2 . Record the best value  𝑎∗1  and  𝑎∗2 
3. Fix  𝑎1=𝑎∗1 , and search  𝑎2  for the best value, call the result the new  𝑎∗2 
4. Fix  𝑎2=𝑎∗2 , and search  𝑎1  for the best value
5. Report the selected  𝑎1  and  𝑎2 
6. Train a model using the selected hyper-parameters, and report the test accuracy 

In [205]:
index = int(np.around(len(x_train)*0.9))
x_subtrain = x_train[0:index]
x_tuning = x_train[index:]
y_subtrain = y_train[0:index]
y_tuning = y_train[index:]

In [211]:
# step 1
tuning = [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 10, 50, 100]
tune_accuracy = []
# step 2
for i in tuning:
    con_binary = np.full((x_subtrain.shape[1],1), i)
    no_reg = np.zeros((1,1))
    lambda_t1 = np.append(con_binary, no_reg, axis = 0)

    logic_t1 = mylogistic_l2(reg_vec = lambda_t1, max_iter = 1000, tol = 1e-5, add_intercept = True)
    logic_t1.fit(x_subtrain, y_subtrain)
    ypred_t1 = logic_t1.predict(x_tuning)
    tune_accuracy.append(accuracy(ypred_t1, y_tuning))

In [212]:
best_coefficient = tuning[np.argmax(tune_accuracy)]
# step 3
tune_a2_acc = []
for i in tuning:
    binary_cnt = 102-6
    numerical = np.full((6,1), best_coefficient)
    binary = np.full((binary_cnt,1), i)
    no_reg = np.zeros((1,1))
    lambda_t2 = np.append(numerical, binary, axis = 0)
    lambda_t2 = np.append(lambda_t2, no_reg, axis = 0)

    logic_t2 = mylogistic_l2(reg_vec = lambda_t2, max_iter = 1000, tol = 1e-5, add_intercept = True)
    logic_t2.fit(x_subtrain, y_subtrain)
    ypred_t2 = logic_t2.predict(x_tuning)
    tune_a2_acc.append(accuracy(ypred_t2, y_tuning))
best_a2 = tuning[np.argmax(tune_a2_acc)]

In [213]:
# step 4
tune_a1_acc = []
for i in tuning:
    binary_cnt = 102-6
    numerical = np.full((6,1), i)
    binary = np.full((binary_cnt,1), best_a2)
    no_reg = np.zeros((1,1))
    lambda_t3 = np.append(numerical, binary, axis = 0)
    lambda_t3 = np.append(lambda_t3, no_reg, axis = 0)

    logic_t3 = mylogistic_l2(reg_vec = lambda_t3, max_iter = 1000, tol = 1e-5, add_intercept = True)
    logic_t3.fit(x_subtrain, y_subtrain)
    ypred_t3 = logic_t3.predict(x_tuning)
    tune_a1_acc.append(accuracy(ypred_t3, y_tuning))
best_a1 = tuning[np.argmax(tune_a1_acc)]
# step 5
print("selected a1:", best_a1)
print("selected a2:", best_a2)

selected a1: 0.001
selected a2: 0.001


In [218]:
# step 6
binary_cnt = 102-6
numerical = np.full((6,1), best_a1)
binary = np.full((binary_cnt,1), best_a2)
no_reg = np.zeros((1,1))
lambda_t4 = np.append(numerical, binary, axis = 0)
lambda_t4 = np.append(lambda_t4, no_reg, axis = 0)

logic_t4 = mylogistic_l2(reg_vec = lambda_t4, max_iter = 1000, tol = 1e-5, add_intercept = True)
logic_t4.fit(x_train, y_train)
ypred_t4 = logic_t4.predict(x_test)
print("test accuracy:", accuracy(ypred_t4, y_test))

test accuracy: 0.8476095617529881


## Q1.5
Use sklearn.linear_model.LogisticRegression to train and test the model (including hyperparameter tuning) <br>
Compare the estimated parameters and test accuracy with those from your own models

In [219]:
from sklearn.linear_model import LogisticRegression

In [239]:
logic_c = [0.001, 0.05, 0.1, 0.5, 1, 5, 10, 50, 100]
sk_tune_acc = []
for i in tuning:
    logic_sk = LogisticRegression(C = i, fit_intercept=True, tol = 1e-5, max_iter = 1000).fit(x_subtrain, y_subtrain.ravel())
    logic_sk.predict(x_tuning)
    sk_tune_acc.append(logic_sk.score(x_tuning, y_tuning.ravel()))
best_sk_coef = tuning[np.argmax(sk_tune_acc)]

logic_sk = LogisticRegression(C = best_sk_coef, fit_intercept=True).fit(x_train, y_train.ravel())
sk_pred = logic_sk.predict(x_test)
test_acc = logic_sk.score(x_test, y_test.ravel())

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [240]:
print("****** my model ******")
print("coefficient for continuous features:", best_a1)
print("coefficient for binary features:", best_a2)
print("test accuracy:", accuracy(ypred_t4, y_test))
print("****** sklearn model ******")
print("coefficient:", best_sk_coef)
print("test accuracy:", test_acc)

****** my model ******
coefficient for continuous features: 0.001
coefficient for binary features: 0.001
test accuracy: 0.8476095617529881
****** sklearn model ******
coefficient: 0.01
test accuracy: 0.7932934926958831


The best parameter $\lambda$ for sklearn model is 0.01, while the best parameter for continuous features and binary features for my own model are both 0.001. <br>
The test accuracy of my own model(0.8476) is better than the test accuracy of sklearn model(0.7933).